### ISSUE TITLE
Recent switch to Saxon 9.8 generates error in fo conversion

### ISSUE THREAD

**Issue by martindholmes** [TEIC/Stylesheets#302]
The xsl:fo conversion of the TEI Simple Print document in the Test2 test suite fails following the recent upgrade of the Saxon version in the lib folder to 9.8. The failure occurs when evaluating the template on line 391 of common/verbatim.xsl (`<xsl:template match="*" mode="verbatim">`). I haven't figured out why yet, but it's an array out-of-bounds error which may be due to one of the many uses of e.g. `preceding-sibling::node()[1]` without first checking if there is a `preceding-sibling::node()`.

To reproduce the error, cd into Test2 and run `ant teiFilesToFo`.

Assigning to myself for the moment, to see if I can figure it out.

**Comment by martindholmes** [comment]
This is a debugging trace leading up to the error. It happens when processing simplePrint.doc.xml:

~~~~~
[java]       <source node="/TEI/text[1]/body[1]/div[17]/div[1]/div[123]/table[1]/row[3]/cell[2]/table[1]/row[2]/cell[2]/table[1]/row[3]/cell[1]/egXML[1]/rendition[1]/namespace::xml" line="3137" file="tei_simplePrint.doc.xml">
[java]        <xsl:variable name="ns" line="756" column="43" module="verbatim.xsl">
[java]         <xsl:choose line="757" column="19" module="verbatim.xsl">
[java]         </xsl:choose>
[java]        </xsl:variable>
[java]       </source><!-- /TEI/text[1]/body[1]/div[17]/div[1]/div[123]/table[1]/row[3]/cell[2]/table[1]/row[2]/cell[2]/table[1]/row[3]/cell[1]/egXML[1]/rendition[1]/namespace::xml -->
[java]      </xsl:for-each>
[java]      <xsl:apply-templates line="770" column="37" module="verbatim.xsl">
[java]       <source node="/TEI/text[1]/body[1]/div[17]/div[1]/div[123]/table[1]/row[3]/cell[2]/table[1]/row[2]/cell[2]/table[1]/row[3]/cell[1]/egXML[1]/rendition[1]/text()[1]" line="3137" file="tei_simplePrint.doc.xml">
[java]        <xsl:template mode="ns" match="text()|comment()|processing-instruction()" line="753" column="78" module="verbatim.xsl">
[java]        </xsl:template>
[java]       </source><!-- /TEI/text[1]/body[1]/div[17]/div[1]/div[123]/table[1]/row[3]/cell[2]/table[1]/row[2]/cell[2]/table[1]/row[3]/cell[1]/egXML[1]/rendition[1]/text()[1] -->
[java]      </xsl:apply-templates>
[java]     </xsl:template>
[java]    </source><!-- /TEI/text[1]/body[1]/div[17]/div[1]/div[123]/table[1]/row[3]/cell[2]/table[1]/row[2]/cell[2]/table[1]/row[3]/cell[1]/egXML[1]/rendition[1] -->
[java]   </xsl:apply-templates>
[java]  </LRE>
[java]  <LRE name="text" line="778" column="15" module="verbatim.xsl">
[java]   <xsl:copy-of line="779" column="36" module="verbatim.xsl">
[java]   </xsl:copy-of>
[java]  </LRE>
[java] </LRE>
[java] <xsl:for-each line="784" column="42" module="verbatim.xsl">
[java]  <source node="/all/names[1]" line="776" file="">
[java]   <xsl:for-each-group line="785" column="56" module="verbatim.xsl">
[java]   </xsl:for-each-group>
[java]  </source><!-- /all/names[1] -->
[java] </xsl:for-each>
[java]</xsl:variable>
[java]              </xsl:template>
[java]             </xsl:call-template>
[java]            </xsl:choose>
[java]           </xsl:if>
[java]           <xsl:choose line="460" column="17" module="verbatim.xsl">
[java]            <xsl:when line="461" column="38" module="verbatim.xsl">
[java]             <xsl:call-template line="462" column="43" module="verbatim.xsl">
[java]              <xsl:copy-of line="464" column="47" module="verbatim.xsl">
[java]              </xsl:copy-of>
[java]              <xsl:text line="465" column="23" module="verbatim.xsl">
[java]              </xsl:text>
[java]              <xsl:template name="Element" line="727" column="32" module="fo_param.xsl">
[java]<xsl:param name="content" line="728" column="32" module="fo_param.xsl">
[java]</xsl:param>
[java]<LRE name="inline" line="729" column="13" module="fo_param.xsl">
[java] <xsl:copy-of line="730" column="39" module="fo_param.xsl">
~~~~~

**Comment by martindholmes** [comment]
It looks like the error comes when it's processing the documentation for the `@scope` attribute on the `<rendition>` element:

~~~~~
<row>
                                            <cell rend="odd_label">scope</cell>
                                            <cell rend="odd_value"><seg xml:lang="en">where CSS is
                                                  used, provides a way of defining <soCalled
                                                  xmlns:rng="http://relaxng.org/ns/structure/1.0"
                                                  xmlns:teix="http://www.tei-c.org/ns/Examples"
                                                  >pseudo-elements</soCalled>, that is, styling
                                                  rules applicable to specific sub-portions of an
                                                  element.</seg><table rend="attDef">
                                                  <row>
                                                  <cell rend="odd_label"><hi rend="label"
                                                  xml:lang="en">Status</hi>
                                                  </cell>
                                                  <cell rend="odd_value"><seg xml:lang="en"
                                                  >Optional</seg></cell>
                                                  </row>
~~~~~

**Comment by sydb** [comment]
@martindholmes and I just narrowed down the problem somewhat, and noticed that there is an `<xsl:copy-of>` on line 730 of fo/fo_param.xsl, but the variable it is outputting (`$content`) is passed as a parameter, and is *always* a string. So I just changed that `<copy-of>` to a `<value-of>` and it worked without bombing. We do not know that it gives the right output, yet.

On a quick check, the output (PDF) looks bad, but no characters are missing or anything.

So @martindholmes is about to push that change.

**Comment by martindholmes** [comment]
Fix committed in #3048467.

**Comment by martindholmes** [comment]
The fix eliminates some nested `<inline>` elements. These don't appear to have been doing anything, since they have no styling attached to them, but I think the ultimate goal might have been to apply styles to these components of the rendered egXML element. I think this needs a little more thought. The situation is that a sequence of strings and elements, of which the last is a string containing a closing angle bracket, is passed in as a parameter to the Element template; that template was echoing them out with xsl:copy-of, but Saxon 9.8 balks at that. Changing to xsl:value-of stringifies them all, which in current situation makes no difference to the output, but it's an undesirable simplification, and it doesn't help us figure out what's actually going on. I'm going to revert the change and try to find out what's really happening with Saxon.

**Comment by martindholmes** [comment]
I cannot figure this out at all. It seems to be a bug in Saxon; as far as I can tell, there's nothing whatsoever wrong with our code, and I can't find a workaround that doesn't trigger the bug. I'm going to stick with the current change, and regenerate the fo results for the current Test suite for the moment.

**Comment by sydb** [comment]
@martindholmes — I’m confused about two things. First:

> The fix eliminates some nested <inline> elements.

It **cannot** change the number of nested elements unless the passed parameter contains element nodes, i.e. is not a string. I thought we’d convinced ourselves it was a string. So maybe other calls to it are not strings?

In either case, I want to verify that you are sure the template that is being called from line 462 of common/verbatim.xsl is the one in fo/fo_param.xsl, as opposed to the one in common/verbatim.xsl (line 800).

Second: when you say “stick with the current change”, do you mean the change of using `<xsl:value-of>` instead of `<xsl:copy-of>`, right?

**Comment by martindholmes** [comment]
The parameter does pass nested elements. Try running the original Test Makefile with test-fo target. Try it with and without the change we made, and echo out contents of the parameter. You'll see that there are lots of nested inline elements, whose purpose I believe is to be available for syntax-highlighting the components of an egXML, but the styling has never been applied. This is definitely a lossy edit, and I don't like it, but I can't figure out a way around it.

I'm sure that's the template being called.

I've stuck with the lossy change we made, but only so I can get past the bug. I'm hoping I can create an isolated test case and raise a Saxon bug.
